source/Halibut.Tests/BiDirectionalTests.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut.ServiceModel;
+using Halibut.Tests.Support;
+using Microsoft.VisualStudio.Threading;
+using NUnit.Framework;
+
+namespace Halibut.Tests
+{
+    public class BiDirectionalTests : BaseTest
+    {
+        // Setting these local variables to avoid Octopus/Tentacle terminology in the test setup
+        readonly X509Certificate2 _endpointACertificate = Certificates.Octopus;
+        readonly X509Certificate2 _endpointBCertificate = Certificates.TentacleListening;
+        readonly string _endpointAThumbprint = Certificates.OctopusPublicThumbprint;
+        readonly string _endpointBThumbprint = Certificates.TentacleListeningPublicThumbprint;
+       
+        [Test]
+        public async Task ClientCanRespondToServerRequests()
+        {
+            await SetupBiDirectionalPollingClients(async (clientServiceA, clientServiceB) =>
+            {
+                var octopusResponse = await clientServiceA.SayHelloAsync("Hello");
+                var tentacleResponse = await clientServiceB.SayHelloAsync("World");
+
+                octopusResponse.Should().Be("Hello...");
+                tentacleResponse.Should().Be("World...");
+            });
+        }
+
+        [Test]
+        public async Task ConcurrentOpposingCommunicationsDoNotBlock()
+        {
+            await SetupBiDirectionalPollingClients(async (clientServiceA, clientServiceB) =>
+            {
+                MyEchoService.SharedState = 0;
+                var taskA = clientServiceA.BlockWaitingForSharedStateIncrementAsync().WithCancellation(CancellationToken);
```

Comment:
This should prevent random race conditions when we run the tests in parallel thanks!